### ISSUE TITLE
Blocked to show and request an AN ad since Unity Plugin RV/Int did not automatically destroy or invalidate the expired AN ad

### ISSUE THREAD

**Issue by nieny225** [mopub/mopub-unity-sdk#187]
### tl;dr:
In the case of using MoPub Unity SDK to mediate Facebook Audience Network: when AN RV or Int ad is expired (>60min), looks like MoPub would not destroy the old ad instance and provide no api to destroy it. As a result, we cannot request a new ad (since MoPub only allow one instance at a time); and if proceed to show the ad, we will get the error as expected (since it's an expired ad).
This makes us miss the opportunity to show an ad.

Mediation: MoPub(5.15.0)/Unity AudienceNetworkPackage(1.8.3)/Android Audience SDK(6.2.0)

### Findings: 
Based on the investigation, we found the Unity Plugin and adapter has some limitation:

**1. onInvalidate() has not been triggered after 60mins**
According to [here](https://github.com/mopub/mopub-android-mediation/blob/master/FacebookAudienceNetwork/src/main/java/com/mopub/mobileads/FacebookRewardedVideo.java#L78), MoPub has a Runnable which seems will trigger onInvalidate() after 60mins to destroy. However, when I manually changed the system time to 2 hours later, the method hadn't been triggered. 

**2. MoPub Unity SDK's [OnRewardedVideoExpiredEvent](https://github.com/mopub/mopub-unity-sdk/blob/ee33ae47a05bc2677262cf87673dd3a6fed37d81/unity-sample-app/Assets/MoPub/Scripts/MoPubManager.cs#L66) and  [OnInterstitialExpiredEvent](https://github.com/mopub/mopub-unity-sdk/blob/ee33ae47a05bc2677262cf87673dd3a6fed37d81/unity-sample-app/Assets/MoPub/Scripts/MoPubManager.cs#L51) would not be fired**
It might link to the first one. Even though MoPub Unity provides OnRewardedVideoExpiredEvent and OnInterstitialExpiredEvent callback, but looks like none of them is called when AN ad is expired. 

**3. MoPub Unity offers [MoPub.HasRewardedVideo()](https://github.com/mopub/mopub-unity-sdk/blob/master/mopub-android-sdk-unity/src/main/java/com/mopub/unity/MoPubRewardedVideoUnityPlugin.java#L108) api but it doesn't include the logic to check AN's isAdInvalidated()**
The HasRewardedVideo is not mapping to [hasvideoavailable](https://github.com/mopub/mopub-android-mediation/blob/master/FacebookAudienceNetwork/src/main/java/com/mopub/mobileads/FacebookRewardedVideo.java#L176). Publisher has no way to check whether the AN ad is valid or not on Unity. 

**4. MoPub doesn't provide destroy method for RV**
For interstitial, publisher can workaround this situation by calling interstitial destroy function, but RV doesn't have that

### Logs:
1. 17:29:14.497 Loaded the RV ad
2. Adjust the system time to two hours later > not seeing any expired event nor onInvalidate() being called
3. 19:29:18.223  Attempt to load a new ad > MoPub returns "Did not queue rewarded ad request" since ad instance is still there, then fired RewardedVideoLoadedEvent again
4. 19:29:25.658 Show the ad > failed to show ad due to expired
![image](https://user-images.githubusercontent.com/6625982/116679650-5ee71080-a9dd-11eb-9221-7c2460148c93.png)

### Steps to Reproduce:
1. Loaded the RV ad
2. Adjust the system time to two hours later
3. load a new ad 
4. Show the ad

### Expected behavior or potential solutions:
- MoPub's runnable should ensure to call onInvalidate() after 60mins. 
- MoPub could provide a destroy method for RV, same as for interstitial
- MoPub could surface an api to allow to proactively check whether AN's ad is still valid or not

### Workarounds (if applicable)
- For interstitial, we can proactively destroy the ad instance after expired. 
- For RV, we can try to show the expired ad to trigger the error, and after that MoPub will trigger onInvalidate(). However, that waste the impression opportunity.

**Comment by chauduyphanvu** [comment]
@nieny225, I think this is only an issue because changing the system time does not have an influence on the Runnable. Based on https://developer.android.com/reference/android/os/SystemClock, this is what the system time is defined as:

> System.currentTimeMillis() is the standard "wall" clock (time and date) expressing milliseconds since the epoch. The wall clock can be set by the user or the phone network (see setCurrentTimeMillis(long)), so the time may jump backwards or forwards unpredictably. This clock should only be used when correspondence with real-world dates and times is important, such as in a calendar or alarm clock application. Interval or elapsed time measurements should use a different clock. If you are using System.currentTimeMillis(), consider listening to the ACTION_TIME_TICK, ACTION_TIME_CHANGED and ACTION_TIMEZONE_CHANGED Intent broadcasts to find out when the time changes.

whereas the Runnable uses `uptimeMillis`:

> uptimeMillis() is counted in milliseconds since the system was booted. This clock stops when the system enters deep sleep (CPU off, display dark, device waiting for external input), but is not affected by clock scaling, idle, or other power saving mechanisms. This is the basis for most interval timing such as Thread.sleep(millls), Object.wait(millis), and System.nanoTime(). This clock is guaranteed to be monotonic, and is suitable for interval timing when the interval does not span device sleep. Most methods that accept a timestamp value currently expect the uptimeMillis() clock.

When you change the system time to "cheat" the Runnable, it has no effect since it's counting the time since boot. One way to test is to modify the underlying adapter's [timer constant](https://github.com/mopub/mopub-android-mediation/blob/master/FacebookAudienceNetwork/src/main/java/com/mopub/mobileads/FacebookRewardedVideo.java#L51) to something shorter (like a few seconds so you can observe the expiration). When I tested this, I was able to see the ad failed, and I could request for another ad without problems.

Vu Chau
MoPub Team
